src/main/java/org/truffleruby/core/string/TStringCache.java

```diff
@@ -69,20 +70,37 @@ private void register(TruffleString tstring, RubyEncoding encoding) {
         }
     }
 
-    public TruffleString getTString(TruffleString string, RubyEncoding encoding) {
-        return getTString(TStringUtils.getBytesOrCopy(string, encoding), encoding);
+    @TruffleBoundary
+    public TruffleString getTString(TruffleString string, RubyEncoding rubyEncoding) {
+        assert rubyEncoding != null;
+
+        var byteArray = string.getInternalByteArrayUncached(rubyEncoding.tencoding);
+        final TBytesKey key = new TBytesKey(byteArray, TStringUtils.hasImmutableInternalByteArray(string),
+                rubyEncoding);
+
+        return getTString(key);
+    }
+
+    public TruffleString getTString(InternalByteArray byteArray, boolean isImmutable, RubyEncoding rubyEncoding) {
+        assert rubyEncoding != null;
+
+        return getTString(new TBytesKey(byteArray, isImmutable, rubyEncoding));
     }
 
-    @TruffleBoundary
     public TruffleString getTString(byte[] bytes, RubyEncoding rubyEncoding) {
```

Comment:
I really wish we had a `reason` field on `@TruffleBoundary` because it's easy to look at this and say "there's no reason this can't be compiled, so lets do away with the boundary". I've been operating under the assumption that anything that could run without a boundary should and let Truffle's heuristics sort out the rest.

Reply:
Feel free to add a code comment about it. Here it's a case of "no value to PE this code, and worse due to forcing an extra allocation that might not be needed otherwise". We should only PE what can benefit from PE, there is a warmup cost to PE too much code.